### PR TITLE
Initial support for Haiku.

### PR DESCRIPTION
**Pull request by waddlesplash**
There's a new GLInterface, but aside from that, mostly just header changes.
virtualbox_haiku x86_64 _22_02_2017_19_34_38

### CODE REVIEW

File: Source/Core/Common/GL/GLInterface/BGL.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#pragma once
+
+#include <Window.h>
+#include <GLView.h>
+
+#include "Common/GL/GLInterfaceBase.h"
+
+class cInterfaceBGL : public cInterfaceBase
+{
+private:
```
Comment (lioncash) [inline Source/Core/Common/GL/GLInterface/BGL.h]: public interface should come before the private interface.

File: Source/Core/Common/GL/GLInterface/BGL.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <Window.h>
+#include <GLView.h>
+
+#include "Common/GL/GLInterfaceBase.h"
+
+class cInterfaceBGL : public cInterfaceBase
```
Comment (lioncash) [inline Source/Core/Common/GL/GLInterface/BGL.h]: This can be a final class

File: Source/Core/Common/GL/GLInterface/BGL.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+void cInterfaceBGL::Swap()
+{
+  gl->SwapBuffers();
+}
+
+// Create rendering window.
+// Call browser: Core.cpp:EmuThread() > main.cpp:Video_Initialize()
+bool cInterfaceBGL::Create(void* window_handle, bool core)
+{
+  window = reinterpret_cast<BWindow*>(window_handle);
```
Comment (lioncash) [inline Source/Core/Common/GL/GLInterface/BGL.cpp]: static_cast

File: Source/Core/Common/GL/GLInterface/BGL.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+{
+private:
+  BWindow* window;
+  BGLView* gl;
```
Comment (lioncash) [inline Source/Core/Common/GL/GLInterface/BGL.h]: This can be a `std::unique_ptr`.

File: Source/Core/Common/GL/GLInterface/BGL.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+#include <Window.h>
+#include <GLView.h>
+
+#include "Common/GL/GLInterfaceBase.h"
+
+class cInterfaceBGL : public cInterfaceBase
+{
+private:
+  BWindow* window;
```
Comment (lioncash) [inline Source/Core/Common/GL/GLInterface/BGL.h]: These can be forward declared, since they're pointer class members.

Comment (lioncash) [inline Source/Core/Common/GL/GLInterface/BGL.h]: I forgot to also point out that member variables should have an `m_` prefix

File: Source/Core/Common/GL/GLInterface/EGLHaiku.cpp [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+EGLNativeWindowType cInterfaceEGLHaiku::InitializePlatform(EGLNativeWindowType host_window,
+                                                             EGLConfig config)
+{
+  EGLint format;
+  eglGetConfigAttrib(egl_dpy, config, EGL_NATIVE_VISUAL_ID, &format);
+  //ANativeWindow_setBuffersGeometry(host_window, 0, 0, format);
+
+  BWindow* window = (BWindow*)host_window;
```
Comment (lioncash) [inline Source/Core/Common/GL/GLInterface/EGLHaiku.cpp]: reinterpret_cast

File: Source/Core/Common/GL/GLInterface/EGLHaiku.cpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Common/GL/GLInterface/EGLHaiku.h"
+
+#include <Window.h>
+
+EGLDisplay cInterfaceEGLHaiku::OpenDisplay()
+{
+	return eglGetDisplay(EGL_DEFAULT_DISPLAY);
```
Comment (sepalani) [inline Source/Core/Common/GL/GLInterface/EGLHaiku.cpp]: I don't think the lint bot will like that tab.

Comment (waddlesplash) [inline Source/Core/Common/GL/GLInterface/EGLHaiku.cpp]: Probably not, I didn't intend to push that file initially but HdkR wanted it.

File: Source/Core/Common/GL/GLInterface/BGL.h (hunk 1 above)
Comment (waddlesplash) [inline Source/Core/Common/GL/GLInterface/BGL.h]: I copied this code mostly from the AGLInterface. Might want to clean that up then.

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

**Comment by ligfx** [comment]
@dolphin-emu-bot rebuild

File: CMakeLists.txt [hunk 8]
```diff
@@ -399,6 +404,11 @@ if(ANDROID)
 	SET(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
 	SET(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
 	SET(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
+elseif(HAIKU)
+	set(USE_X11 0)
+	set(USE_UPNP 0)
+	set(USE_EGL 0)
+	list(APPEND LIBS network)
```
Comment (Orphis) [inline CMakeLists.txt]: Does EVERYTHING need the network library?
Which target actually requires it? Can't you add the dependency there instead?

File: CMakeLists.txt [hunk 9]
```diff
@@ -438,7 +448,7 @@ endif()
 
 set(USE_X11 0)
 
-if(UNIX AND NOT APPLE AND NOT ANDROID AND NOT ENABLE_HEADLESS)
+if(UNIX AND NOT APPLE AND NOT ANDROID AND NOT HAIKU AND NOT ENABLE_HEADLESS)
```
Comment (Orphis) [inline CMakeLists.txt]: Is X11 available on Haiku?
If so, why don't you want to use it there?
If not, why can't you just let the X11 detection fail? It's fine to do so and let it disable the X11 support then.

File: CMakeLists.txt (hunk 8 above)
Comment (waddlesplash) [inline CMakeLists.txt]: Dolphin itself uses POSIX socket functions in the Core. And the LIBS variable seems to only affect that (see traversal_server which I also had to add a patch for.)

File: CMakeLists.txt (hunk 9 above)
Comment (waddlesplash) [inline CMakeLists.txt]: Haiku has its own display server/window manager which is not even remotely compatible with X11. And if Dolphin can't find X11, you run into lines 465 and following:
```
if (NOT USE_X11)
 		message(FATAL_ERROR "\n"
 			"No suitable display platform found\n"
 			"Requires x11 to run")
endif()
```
So Haiku is basically equivalent to macOS/Android in this regard, and they're also special cased here.

Comment (Orphis) [inline CMakeLists.txt]: I believe that error message could be removed instead then.

File: CMakeLists.txt (hunk 8 above)
Comment (Orphis) [inline CMakeLists.txt]: Then add the dependency to "network" directly in the core if that's where it is needed, don't add it globally.
That LIBS variable needs to be removed and explicit direct dependencies on each target should be added instead.

File: CMakeLists.txt (hunk 9 above)
Comment (waddlesplash) [inline CMakeLists.txt]: I don't think it's my place to do that (and I was trying to keep this PR to a minimum; I already moved some OS-independent changes into another PR which was already merged)

**Comment by waddlesplash** [comment]
All comments & merge conflicts fixed; all tests passed. Merge?

File: CMakeLists.txt (hunk 9 above)
Comment (Orphis) [inline CMakeLists.txt]: I disagree with that. If you know a clean way to fix something, just do it then!

File: Externals/enet/CMakeLists.txt [hunk 10]
```diff
@@ -6,6 +6,11 @@ project(enet)
 include(CheckFunctionExists)
 include(CheckStructHasMember)
 include(CheckTypeSize)
+
+if(HAIKU)
+  set(CMAKE_REQUIRED_LIBRARIES network)
```
Comment (Orphis) [inline Externals/enet/CMakeLists.txt]: Why do you do this when you have a target_link_libraries() below?

File: Source/Core/Common/CMakeLists.txt [hunk 11]
```diff
@@ -120,4 +123,7 @@ add_dolphin_library(common "${SRCS}" "${LIBS}")
 if(UNIX)
   # Posix networking code needs to be fixed for Windows
   add_executable(traversal_server TraversalServer.cpp)
+  if(HAIKU)
+    target_link_libraries(traversal_server network)
```
Comment (Orphis) [inline Source/Core/Common/CMakeLists.txt]: `target_link_libraries(traversal_server PRIVATE network)`

File: Externals/enet/CMakeLists.txt (hunk 10 above)
Comment (Orphis) [inline Externals/enet/CMakeLists.txt]: Oh right, that would be for the check_function_exists calls. It's alright then!

**Comment by endrift** [comment]
You going to add a Haiku buildbot to make sure the build isn't constantly broken?

**Comment by waddlesplash** [comment]
Can someone review/rebuild?

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

**Pull request merged**